keps/sig-auth/0000-20170814-bounding-self-labeling-kubelets.md

```diff
@@ -0,0 +1,163 @@
+---
+kep-number: 0
+title: Bounding Self-Labeling Kubelets
+authors:
+  - "@mikedanese"
+  - "@liggitt"
+owning-sig: sig-auth
+participating-sigs:
+  - sig-node
+  - sig-storage
+reviewers:
+  - "@saad-ali"
+  - "@tallclair"
+  - "@deads2k"
+approvers:
+  - "@tallclair"
+  - "@smarterclayton"
+creation-date: 2017-08-14
+last-updated: 2018-10-25
+status: provisional
+---
+
+# Bounding Self-Labeling Kubelets
+
+## Motivation
+
+Today the node client has total authority over its own Node labels.
+This ability is incredibly useful for the node auto-registration flow.
+The kubelet reports a set of well-known labels, as well as additional
+labels specified on the command line with `--node-labels`.
+
+While this distributed method of registration is convenient and expedient, it
+has two problems that a centralized approach would not have. Minorly, it makes
+management difficult. Instead of configuring labels in a centralized
+place, we must configure `N` kubelet command lines. More significantly, the
+approach greatly compromises security. Below are two straightforward escalations
+on an initially compromised node that exhibit the attack vector.
+
+### Capturing Dedicated Workloads
+
+Suppose company `foo` needs to run an application that deals with PII on
+dedicated nodes to comply with government regulation. A common mechanism for
+implementing dedicated nodes in Kubernetes today is to set a label or taint
+(e.g. `foo/dedicated=customer-info-app`) on the node and to select these
+dedicated nodes in the workload controller running `customer-info-app`.
+
+Since the nodes self reports labels upon registration, an intruder can easily
+register a compromised node with label `foo/dedicated=customer-info-app`. The
+scheduler will then bind `customer-info-app` to the compromised node potentially
+giving the intruder easy access to the PII.
+
+This attack also extends to secrets. Suppose company `foo` runs their outward
+facing nginx on dedicated nodes to reduce exposure to the company's publicly
+trusted server certificates. They use the secret mechanism to distribute the
+serving certificate key. An intruder captures the dedicated nginx workload in
+the same way and can now use the node certificate to read the company's serving
+certificate key.
+
+## Proposal
+
+The NodeRestriction admission plugin will be updated to limit the labels a kubelet
+can set on its Node object. Kubelets will only be allowed to set the following labels.
+
+### Well-known label keys
+
+This list includes all of the labels that kubelets already set on their Node objects.
+
+```
+kubernetes.io/hostname
+failure-domain.beta.kubernetes.io/zone
+failure-domain.beta.kubernetes.io/region
+beta.kubernetes.io/instance-type
+beta.kubernetes.io/os
+beta.kubernetes.io/arch
```

Comment:
so if i understand this correctly , the fix is to NOT allow arbitrary labels from kubelet but only well defines ones ? Currently we specify some pool labels for nodes using kubelet, where will that go now ?

Reply:
> Currently we specify some pool labels for nodes using kubelet, where will that go now?

You would choose labels with a `kubelet.kubernetes.io/` or `topology.kubernetes.io/` prefix to represent your node pool labels you wanted the kubelet to set. See the second and third sections for examples.